### PR TITLE
Adding label/tags to cypht

### PR DESCRIPTION
**Pull request by Shadow243**
Related issue: [https://github.com/cypht-org/cypht/issues/13](https://github.com/cypht-org/cypht/issues/13)

Screenshot 2024-06-04 at 11 58 26

### CODE REVIEW

**Comment by marclaporte** [comment]
This was requested 9 years ago, so I expect it will make a lot of people happy! https://github.com/cypht-org/cypht/issues/13

Can we envisage that when Cypht is integrated with other apps (like Tiki), it shares tags/labels? We don't need now, but I expect we will want later.

**Comment by kroky** [comment]
Looks good so far. I'll review once this is ready out of draft mode.

**Comment by Shadow243** [comment]
> I'll review once this is ready out of draft mode.

Ready for review @kroky

File: modules/core/site.js [hunk 1]
```diff
@@ -1573,6 +1575,19 @@ var Hm_Utils = {
         return res;
     },
 
+    seach_from_local_storage: function(pattern) {
```
Comment (kroky) [inline modules/core/site.js:1578]: Search is misspelled here. It seems rather inefficient to look through all the entries in sessionStorage. Also, save_to_local_storage checks for Storage != void(0) to prevent other JS errors, we should do the same here, I think.

What is the purpose of this function? You seem to load all the stored imap source keys in the storage and save back with value of 1. Not sure I understand the need to do this. Do you invalidate some kind of caching?

File: modules/imap/functions.php [hunk 2]
```diff
@@ -1321,6 +1321,49 @@ function snooze_message($imap, $msg_id, $folder, $snooze_tag) {
...
+    }
+    $msg = "X-Cypht-Labels:".implode(',', $tags)."\n".$msg;
+
+    $msg = str_replace("\r\n", "\n", $msg);
+    $msg = str_replace("\n", "\r\n", $msg);
+    $msg = rtrim($msg)."\r\n";
+
+    $res = false;
+
+    if (!count($imap->get_mailbox_status($folder))) {
```
Comment (kroky) [inline modules/imap/functions.php:1352]: If you just downloaded a message from the folder it seems impossible for the folder to not exist and create it here. Better leave the folder checking at the top and if it doesn't exist return early. Otherwise, just use the folder. We can't create a folder to download a message from it.

File: modules/imap/functions.php [hunk 3]
```diff
@@ -1321,6 +1321,49 @@ function snooze_message($imap, $msg_id, $folder, $snooze_tag) {
...
+        $imap->create_mailbox($folder);
+    }
+    if ($imap->select_mailbox($folder) && $imap->append_start($folder, strlen($msg))) {
+        $imap->append_feed($msg."\r\n");
+        if ($imap->append_end()) {
+            if ($imap->select_mailbox($folder) && $imap->message_action('DELETE', array($msg_id))) {
```
Comment (kroky) [inline modules/imap/functions.php:1358]: I think you should only select the mailbox once at the top of the method. No need to further select the same folder as you are working in one folder only.

File: modules/imap/functions.php [hunk 4]
```diff
@@ -1417,6 +1460,27 @@ function snooze_dropdown($output, $unsnooze = false) {
     return $txt;
 }}
 
+if (!hm_exists('tags_dropdown')) {
+function tags_dropdown($context, $headers) {
+    $folders = $context->get('tag_folders', array());
```
Comment (kroky) [inline modules/imap/functions.php:1465]: These seem to be tag names or labels, why do we call them tag folders? Better call them just labels or tags.

File: modules/tags/functions.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
+<?php
+
+use PhpParser\Node\Expr\AssignOp\Div;
```
Comment (kroky) [inline modules/tags/functions.php:3]: Where is this used here?

File: modules/tags/site.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    //     Hm_Ajax.request(
+    //         form.serializeArray(),
+    //         function(res) {
+    //             console.log("res", res)
+    //             // if (res.) {
+    //             // }
+    //         }
+    //     );
+        
+    // });
```
Comment (kroky) [inline modules/tags/site.js:16]: This seems like debugging code - can you please remove it?

File: modules/core/site.js (hunk 1 above)
Comment (Shadow243) [reply modules/core/site.js:1578]: > Do you invalidate some kind of caching?

Yes exactly, when I access the message page a cache like: 62_imap_xxxxxxxx_xxxx is created. The problem is that the next time when I open this email again it will be retrieved from the cache so if I add a label I won't be able to get it updated. This is why I added a function that allows me to search the cache for anything that has this format: int_imap_xxxxxxxx_xxxx and delete them

File: modules/imap/functions.php (hunk 2 above)
Comment (Shadow243) [reply modules/imap/functions.php]: Done

File: modules/imap/functions.php (hunk 3 above)
Comment (Shadow243) [reply modules/imap/functions.php]: Removed

File: modules/imap/functions.php (hunk 4 above)
Comment (Shadow243) [reply modules/imap/functions.php]: Done, i'm now calling them `labels`

File: modules/tags/functions.php (hunk 5 above)
Comment (Shadow243) [reply modules/tags/functions.php]: Removed

**Comment by marclaporte** [comment]
Reading the code, I see both tag and label being used. Please standardize to one.

It's OK to have code comment which indicates the alternative name for searchability.

**Comment by Shadow243** [comment]
> Please standardize to one.

@marclaporte I'm now using tag/tags.
